v2/tools/generator/internal/codegen/debug_reporter.go

```diff
@@ -0,0 +1,221 @@
+// groupSelector specifies which groups to include (may include wildcards).
+// outputFolder specifies where to write the debug output.
+func newDebugReporter(groupSelector string, outputFolder string) *debugReporter {
+	return &debugReporter{
+		groupSelector: config.NewStringMatcher(groupSelector),
+		outputFolder:  outputFolder,
+	}
+}
+
+func (dr *debugReporter) ReportStage(stage int, description string, state *pipeline.State) error {
+	if dr == nil || dr.outputFolder == "" {
+		// Not in debug mode.
+		return nil
+	}
+
+	report, err := dr.createReport(state)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create debug report for stage %d", stage)
+	}
+
+	// Save the template to the debug directory
+	filename := dr.createFileName(stage, description)
+	err = ioutil.WriteFile(filename, []byte(report), 0600)
+	return errors.Wrapf(err, "failed to write debug log to %s", filename)
+}
+
+func (dr *debugReporter) createReport(state *pipeline.State) (string, error) {
+	var buffer strings.Builder
+
+	packages := dr.findPackages(state.Definitions())
+	for _, pkg := range packages {
+		grp, _ := pkg.GroupVersion()
+		if !dr.groupSelector.Matches(grp) {
+			// Skip this package
+			continue
+		}
+
+		rpt := newDebugReport(pkg.PackagePath())
+		dr.writeDefinitions(rpt, dr.inPackage(pkg, state.Definitions()))
+
+		err := rpt.saveTo(&buffer)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to create debug report for %s", pkg.PackagePath())
+		}
+	}
+
+	return buffer.String(), nil
+}
+
+// writeDefinitions writes the definitions to the debug report.
+// rpt is the debug report to write to.
+// definitions is the set of definitions to write.
+// Definitions are written in alphabetical order.
+func (dr *debugReporter) writeDefinitions(rpt *debugReport, definitions astmodel.TypeDefinitionSet) {
+	defs := definitions.AsSlice()
+	sort.Slice(defs, func(i, j int) bool {
+		return defs[i].Name().Name() < defs[j].Name().Name()
+	})
+
+	for _, d := range defs {
+		dr.writeDefinition(rpt, d)
+	}
+}
+
+// writeDefinition writes the definition to the debug report.
+// rpt is the debug report to write to.
+// definition is the definition to write.
+func (dr *debugReporter) writeDefinition(rpt *debugReport, definition astmodel.TypeDefinition) {
+	name := definition.Name()
+	t := definition.Type()
+	sub := rpt.add(
+		fmt.Sprintf("%s: %s", name.Name(), astmodel.DebugDescription(t, name.PackageReference)))
+	dr.writeType(sub, definition.Type(), name.PackageReference)
+}
+
+// writeType writes the type to the debug report.
+// rpt is the debug report to write to.
+// t is the type to write.
+// currentPackage is the package that the type is defined in (used to simplify type descriptions).
+// Only complex types where astmodel.DebugDescription is insufficient are written.
+func (dr *debugReporter) writeType(
+	rpt *debugReport,
+	t astmodel.Type,
+	currentPackage astmodel.PackageReference,
+) {
+	if rsrc, ok := astmodel.AsResourceType(t); ok {
```

Comment:
I initiallly prototyped this with a TypeVisitor, but it was both slow and didn't quite do the iteration in the way that I wanted (I wanted tight control to influence the format of the generated debug output), so I went with the recursive method calls approach instead. This is also slightly faster, which is useful since we dump a debug report after every stage.